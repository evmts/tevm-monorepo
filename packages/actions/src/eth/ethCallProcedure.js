import { callProcedure } from '../call/callProcedure.js'

/**
 * Executes a message call without creating a transaction on the block chain.
 * @param {import('@tevm/node').TevmNode} client
 * @returns {import('./EthProcedure.js').EthCallJsonRpcProcedure}
 */
export const ethCallProcedure = (client) => async (req) => {
	const [tx, blockTag, stateOverrideSet, blockOverrideSet] = req.params
	const { data, from, to, gas, gasPrice, value } = tx
	const response = await callProcedure(client)({
		...(req.id !== undefined ? { id: req.id } : {}),
		jsonrpc: req.jsonrpc,
		method: 'tevm_call',
		params: [
			{
				...(gasPrice !== undefined ? { gasPrice } : {}),
				...(data !== undefined ? { data } : {}),
				...(gas !== undefined ? { gas } : {}),
				...(value !== undefined ? { value } : {}),
				...(to !== undefined ? { to } : {}),
				...(from !== undefined ? { from } : {}),
				...(blockTag !== undefined ? { blockTag } : {}),
			},
			stateOverrideSet,
			blockOverrideSet,
		],
	})
	if (!response.result) {
		return {
			jsonrpc: req.jsonrpc,
			method: 'eth_call',
			error: response.error,
			...(req.id !== undefined ? { id: req.id } : {}),
		}
	}
	return {
		jsonrpc: req.jsonrpc,
		method: 'eth_call',
		result: response.result.rawData,
		...(req.id !== undefined ? { id: req.id } : {}),
	}
}
