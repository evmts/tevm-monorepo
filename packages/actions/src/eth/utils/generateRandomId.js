/**
 * @returns {import("@tevm/utils").Hex}
 */
export const generateRandomId = () => {
	return `0x${Array.from(crypto.getRandomValues(new Uint8Array(16)))
		.map((b) => b.toString(16).padStart(2, '0'))
		.join('')}`
}
