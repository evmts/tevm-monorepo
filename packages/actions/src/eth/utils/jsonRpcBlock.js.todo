import { bytesToHex } from "viem"

const ethjsTxToTevmTx = (tx) => {

}

/**
 * Transoforms an ethjs block to a tevm Block
 * @param {import("@ethereumjs/block").Block} block
 * @param {boolean} includeTransactions
 * @returns {Promise<import("@tevm/actions-types").Block>}
 */
const ethjsToTevmBlock = async (
  block,
  includeTransactions,
) => {
  const transactions = block.transactions.map((tx, txIndex) =>
    includeTransactions ? ethjsTxToTevmTx(tx, block, txIndex) : bytesToHex(tx.hash())
  )
  const withdrawalsAttr =
    block.header.withdrawalsRoot !== undefined
      ? {
        withdrawalsRoot: bytesToHex(block.header.withdrawalsRoot),
        withdrawals: block.toJSON().withdrawals,
      }
      : {}
  const td = await chain.getTd(block.hash(), block.header.number)
  return {
    coinbase: block.toJSON().header?.coinbase
    number: block.toJSON().header.number,
    hash: bytesToHex(block.hash()),
    parentHash: block.toJSON().header.parentHash,
    mixHash: block.toJSON().header.mixHash,
    nonce: block.toJSON().header.nonce,
    sha3Uncles: block.toJSON().header.uncleHash,
    logsBloom: block.toJSON().header.logsBloom,
    transactionsRoot: block.toJSON().header.transactionsTrie,
    stateRoot: block.toJSON().header.stateRoot,
    receiptsRoot: block.toJSON().header.receiptTrie,
    miner: block.toJSON().header.coinbase,
    difficulty: block.toJSON().header.difficulty,
    totalDifficulty: bigIntToHex(td),
    extraData: block.toJSON().header.extraData,
    size: intToHex(utf8ToBytes(JSON.stringify(block.toJSON())).byteLength),
    gasLimit: block.toJSON().header.gasLimit,
    gasUsed: block.toJSON().header.gasUsed,
    timestamp: block.toJSON().header.timestamp,
    transactions,
    uncles: block.uncleHeaders.map((uh) => bytesToHex(uh.hash())),
    baseFeePerGas: block.toJSON().header.baseFeePerGas,
    ...withdrawalsAttr,
    blobGasUsed: block.toJSON().header.blobGasUsed,
    excessBlobGas: block.toJSON().header.excessBlobGas,
    parentBeaconBlockRoot: block.toJSON().header.parentBeaconBlockRoot,
  }
}
