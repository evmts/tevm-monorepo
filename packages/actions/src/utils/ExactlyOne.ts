/**
 * Helper type to ensure exactly one property from a set is provided
 */
export type ExactlyOne<T, K extends keyof T> = {
	[P in K]: { [Q in P]: T[P] } & { [Q in Exclude<K, P>]?: never }
}[K]
