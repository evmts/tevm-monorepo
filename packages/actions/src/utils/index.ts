export * from './SerializeToJson.js'
export * from './blockToJsonRpcBlock.js'
export * from './generateRandomId.js'
export * from './parseBlockTag.js'
export * from './txToJSONRPCTx.js'
