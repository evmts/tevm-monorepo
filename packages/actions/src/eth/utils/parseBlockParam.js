import { InvalidBlockError } from '@tevm/errors'
import { hexToBigInt, hexToBytes } from '@tevm/utils'

/**
 * @param {import('@tevm/blockchain').Chain} blockchain
 * @param {import('../../common/BlockParam.js').BlockParam} blockParam
 * @returns {Promise<bigint >}
 */
export const parseBlockParam = async (blockchain, blockParam) => {
	if (typeof blockParam === 'number') {
		return BigInt(blockParam)
	}
	if (typeof blockParam === 'bigint') {
		return blockParam
	}
	if (typeof blockParam === 'string' && blockParam.startsWith('0x')) {
		if (blockParam.length === 66) {
			const block = await blockchain.getBlock(hexToBytes(/** @type {import('@tevm/utils').Hex}*/ (blockParam)))
			return BigInt(block.header.number)
		}
		return hexToBigInt(/** @type {import('@tevm/utils').Hex}*/ (blockParam))
	}
	if (blockParam === 'safe') {
		const safeBlock = blockchain.blocksByTag.get('safe')
		// let's handle it here in case we forget to update it later
		if (safeBlock) {
			return safeBlock.header.number
		}
		throw new InvalidBlockError('safe not currently supported as block tag')
	}
	if (blockParam === 'latest' || blockParam === undefined) {
		const safeBlock = blockchain.blocksByTag.get('latest')
		// let's handle it here in case we forget to update it later
		if (safeBlock) {
			return safeBlock.header.number
		}
		throw new InvalidBlockError('latest block does not exist on chain')
	}
	if (blockParam === 'pending') {
		// for pending we need to mine a new block and then handle it
		// let's skip this functionality for now
		throw new InvalidBlockError(
			'Pending not yet supported but will be in future. Consider opening an issue or reaching out on telegram if you need this feature to expediate its release',
		)
	}
	if (blockParam === 'earliest') {
		return BigInt(1)
	}
	if (blockParam === 'finalized') {
		throw new InvalidBlockError('finalized noet yet supported for this feature')
	}
	blockchain.logger.error({ blockParam }, 'Unknown block param pased to blockNumberHandler')
	throw new InvalidBlockError(`Unknown block param ${blockParam} pased to blockNumberHandler`)
}
