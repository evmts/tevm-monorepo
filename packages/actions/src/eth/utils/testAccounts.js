import { mnemonicToAccount } from '@tevm/utils'

const oneToTen = Array.from(Array(10).keys())

export const mnemonic =
	'test test test test test test test test test test test junk'

export const testAccounts =
	/**
	 * @type {[import("@tevm/utils").HDAccount, import("@tevm/utils").HDAccount,import("@tevm/utils").HDAccount, import("@tevm/utils").HDAccount,import("@tevm/utils").HDAccount, import("@tevm/utils").HDAccount,import("@tevm/utils").HDAccount, import("@tevm/utils").HDAccount,import("@tevm/utils").HDAccount, import("@tevm/utils").HDAccount]}
	 */
	(oneToTen.map((i) => mnemonicToAccount(mnemonic, { addressIndex: i })))
