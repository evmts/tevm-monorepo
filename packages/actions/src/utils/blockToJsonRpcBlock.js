import { bytesToHex, numberToHex, toBytes } from '@tevm/utils'
import { txToJSONRPCTx } from './txToJSONRPCTx.js'

/**
 * @param {import('@tevm/block').Block} block
 * @param {boolean} includeTransactions
 * @returns {Promise<Required<import('../eth/EthJsonRpcResponse.js').EthGetBlockByHashJsonRpcResponse>['result']>}
 */
export const blockToJsonRpcBlock = async (block, includeTransactions) => {
	const json = block.toJSON()
	const header = /** @type {import('@tevm/block').JsonHeader}*/ (json.header)
	const transactions = block.transactions.map((tx, txIndex) =>
		includeTransactions ? txToJSONRPCTx(tx, block, txIndex) : bytesToHex(tx.hash()),
	)

	/**
	 * @type {import('../eth/EthJsonRpcResponse.js').EthGetBlockByHashJsonRpcResponse['result']}
	 */
	const out = {
		number: /** @type {import('@tevm/utils').Hex}*/ (header.number),
		hash: bytesToHex(block.hash()),
		parentHash: /** @type {import('@tevm/utils').Hex}*/ (header.parentHash),
		// TODO add this to the type
		...{ mixHash: header.mixHash },
		nonce: /** @type {import('@tevm/utils').Hex}*/ (header.nonce),
		sha3Uncles: /** @type {import('@tevm/utils').Hex}*/ (header.uncleHash),
		logsBloom: /** @type {import('@tevm/utils').Hex}*/ (header.logsBloom),
		transactionsRoot: /** @type {import('@tevm/utils').Hex}*/ (header.transactionsTrie),
		stateRoot: /** @type {import('@tevm/utils').Hex}*/ (header.stateRoot),
		miner: /** @type {import('@tevm/utils').Address}*/ (header.coinbase),
		difficulty: /** @type {import('@tevm/utils').Hex}*/ (header.difficulty),
		// TODO we need to actually add this
		totalDifficulty: /** @type {import('@tevm/utils').Hex}*/ ('0x0'),
		extraData: /** @type {import('@tevm/utils').Hex}*/ (header.extraData),
		size: numberToHex(toBytes(JSON.stringify(json)).byteLength),
		gasLimit: /** @type {import('@tevm/utils').Hex}*/ (header.gasLimit),
		gasUsed: /** @type {import('@tevm/utils').Hex}*/ (header.gasUsed),
		timestamp: /** @type {import('@tevm/utils').Hex}*/ (header.timestamp),
		uncles: block.uncleHeaders.map((uh) => bytesToHex(uh.hash())),
		// TODO fix this type
		transactions: /** @type any*/ (transactions),
		// TODO add this to the type
		...{ baseFeePerGas: header.baseFeePerGas },
		...{ receiptsRoot: header.receiptTrie },
		...(header.withdrawalsRoot !== undefined
			? {
					withdrawalsRoot: header.withdrawalsRoot,
					withdrawals: json.withdrawals,
				}
			: {}),
		...(header.blobGasUsed !== undefined ? { blobGasUsed: header.blobGasUsed } : {}),
		// TODO add this to the type
		...{ requestsRoot: header.requestsRoot },
		// TODO add this to the type
		...{ requests: block.requests?.map((req) => bytesToHex(req.serialize())) },
		// TODO add this to the type
		...{ excessBlobGas: header.excessBlobGas },
		// TODO add this to the type
		...{ parentBeaconBlockRoot: header.parentBeaconBlockRoot },
	}

	return out
}
