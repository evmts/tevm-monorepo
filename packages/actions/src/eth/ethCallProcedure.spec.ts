import { createTevmNode, type TevmNode } from '@tevm/node'
import { PREFUNDED_ACCOUNTS } from '@tevm/utils'
import { beforeEach, describe, expect, it } from 'vitest'
import type { EthCallJsonRpcRequest } from './EthJsonRpcRequest.js'
import { ethCallProcedure } from './ethCallProcedure.js'

let client: TevmNode

beforeEach(() => {
	client = createTevmNode()
})

describe('ethCallProcedure', () => {
	it('should execute a message call successfully', async () => {
		const request: EthCallJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_call',
			id: 1,
			params: [
				{
					data: '0x0',
					from: PREFUNDED_ACCOUNTS[0].address,
					to: `0x${'69'.repeat(20)}`,
					value: '0x1',
				},
				'latest',
			],
		}

		const response = await ethCallProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_call')
		expect(response.id).toBe(request.id as any)
		expect(response.result).toMatchSnapshot()
	})

	it('should handle missing optional fields in the request', async () => {
		const request: EthCallJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_call',
			id: 1,
			params: [
				{
					from: PREFUNDED_ACCOUNTS[0].address,
					to: `0x${'69'.repeat(20)}`,
				},
				'latest',
			],
		}

		const response = await ethCallProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_call')
		expect(response.id).toBe(request.id as any)
		expect(response.result).toMatchSnapshot()
	})

	it('should handle requests without an id', async () => {
		const request: EthCallJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_call',
			params: [
				{
					from: PREFUNDED_ACCOUNTS[0].address,
					to: `0x${'69'.repeat(20)}`,
				},
				'latest',
			],
		}

		const response = await ethCallProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_call')
		expect(response.id).toBeUndefined()
		expect(response.result).toMatchSnapshot()
	})

	it('should not create transactions or trigger mining during call', async () => {
		const automineClient = createTevmNode({ miningConfig: { type: 'auto' } })
		const initialBlock = await automineClient.getVm().then((vm) => vm.blockchain.getCanonicalHeadBlock())
		const initialBlockNumber = initialBlock.header.number

		const txPool = await automineClient.getTxPool()
		const initialTxPoolSize = txPool.pool.size

		const request: EthCallJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_call',
			id: 1,
			params: [
				{
					from: PREFUNDED_ACCOUNTS[0].address,
					to: `0x${'69'.repeat(20)}`,
				},
				'latest',
			],
		}

		const response = await ethCallProcedure(automineClient)(request)

		const finalBlock = await automineClient.getVm().then((vm) => vm.blockchain.getCanonicalHeadBlock())
		const finalBlockNumber = finalBlock.header.number
		const finalTxPoolSize = txPool.pool.size

		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(finalBlockNumber).toBe(initialBlockNumber)
		expect(finalTxPoolSize).toBe(initialTxPoolSize)
	})
})
