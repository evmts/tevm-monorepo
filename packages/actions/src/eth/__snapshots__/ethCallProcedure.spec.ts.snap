// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ethCallProcedure > should execute a message call successfully 1`] = `"0x"`;

exports[`ethCallProcedure > should handle missing optional fields in the request 1`] = `"0x"`;

exports[`ethCallProcedure > should handle requests without an id 1`] = `"0x"`;

exports[`ethCallProcedure should execute a message call successfully 1`] = `"0x"`;

exports[`ethCallProcedure should handle missing optional fields in the request 1`] = `"0x"`;

exports[`ethCallProcedure should handle requests without an id 1`] = `"0x"`;
